### ISSUE TITLE
How stable is the API right now? 

### ISSUE THREAD

**Issue by RedBearAK** [ulinja/hyprpy#10]
@ulinja 

I was looking for a way to get some simple things like the active window class and title from Python, for use with a keymapper, and I found your project on PyPI. 

Looks simpler to use, and doesn't seem to use `async` functions, which kind of made `hyprland-py` incompatible with the original project I'm working with. 

Question is, how stable do you think the API is at this stage? I see that the repo is only about a month old. If I integrate this as part of my keymapper solution for getting window context on Hyprland, how likely do you think it would be that I'd need to modify my own code later? Is there still some initial flux happening? 

And another question is whether using this would be any different (efficiency-wise) from an existing solution that uses the Unix socket that accepts `hyprctl` style queries. Your project uses the sockets that Hyprland makes available, if I understand correctly.

**Comment by ulinja** [comment]
Hi @RedBearAK

that's a great question! I don't want to make false promises, there is some potential for the API to change.
At its current stage, with simple event handling and read access to Component properties, the library provides all the functionality that I use _personally_ at this point. However, there is much room for adding functionality, such as [Hyprland dispatchers](https://wiki.hyprland.org/Configuring/Dispatchers/). These allow implementing things like calling `Window.toggle_fullscreen()` to modify Components (toggling a Window's fullscreen state in this case).
I don't have a specific roadmap for this yet, it depends on whether there are any feature requests for this kind of thing.

What I _can_ promise is that the API for read access to [simple Component properties](https://hyprpy.lobbes.dev/hyprpy.api-reference.data.html), things like `Window.wm_class` will not change. These are all properties pulled directly from the Hyprland socket, and their names will not change in the future. The API for access to related components will also stay the same.

Of course, I will do my best to keep changes for added functionality backwards compatible.

As for your second question, `hyprctl` [is a simple C++ wrapper for writing to the Hyprland socket](https://github.com/hyprwm/Hyprland/blob/main/hyprctl/main.cpp).
[Hyprpy does exactly the same thing](https://github.com/ulinja/hyprpy/blob/fa4a44197d41825e6847a322952d1a474967e8aa/hyprpy/utils/sockets.py#L184), and uses Python's standardlib sockets library, which in turn is written in C.
However, here are some performance differences to consider:
- `hyprctl` spawns a new process each time you invoke it from your shell, and returns text for you to parse yourself. `hyprpy` does not spawn new processes when talking to the socket, and takes care of parsing into Python objects for you.
- `hyprpy` uses Pydantic, a parsing library written in Rust, to parse socket responses, but parsing and instantiating Python objects does present a performance overhead.

Overall, the performance difference in socket operations is not that great. Using Hyprpy signals to avoid polling, and to avoid spawning/killing hyprctl processes repeatedly _can_ make a big difference. As usual, it depends on your usecase.

**Comment by RedBearAK** [comment]
The fact that you took the time to write that up gives me good feelings. Thanks for the response. 

But I got impatient when you didn't reply within 36 nanoseconds, so I went ahead and integrated `hyprpy` as one of three cascading (potential, untested) solutions for grabbing the window class and title in Hyprland, for use with a keymapper doing app-specific remaps. 

https://github.com/RedBearAK/keyszer/blob/3fe6e5213b8c3a956dd404d2038cc0a917ce8423/src/keyszer/lib/window_context.py#L85

The other methods are A) working directly with the socket with the built-in Python options, and then B) the much slower use of `hyprctl` shell commands, which will probably never get used, even as a last resort. 

Unfortunately I don't use Hyprland, so I can't actually test any of this yet, but I have a user interested in my project that should be able to test it soon. 

All I really need is to be able to reliably grab the window class/title, so I'm not too worried about needing to update the code later. 

The method that uses `hyprpy` is certainly more compact than the other methods, so that's nice.
